stdlib/Dates/src/periods.jl

```diff
@@ -201,6 +205,26 @@ CompoundPeriod(t::Time) = CompoundPeriod(Period[Hour(t), Minute(t), Second(t), M
 
 CompoundPeriod(p::Period...) = CompoundPeriod(Period[p...])
 
+time_per_time(::Type{P}, ::Type{P}) where {P} = 1
+function define_time_per_times(periods)
+    for i = eachindex(periods)
+        T, n = periods[i]
+        N = Int64(1)
+        for j = (i - 1):-1:firstindex(periods) # less-precise periods
+            Tc, nc = periods[j]
+            N *= nc
+            @eval time_per_time(::Type{$T}, ::Type{$Tc}) = $N
+        end
+    end
+end
+
+define_time_per_times([(:Week, 7), (:Day, 24), (:Hour, 60), (:Minute, 60), (:Second, 1000),
+                    (:Millisecond, 1000), (:Microsecond, 1000), (:Nanosecond, 1)])
+
+function CompoundPeriod(x::Real, ::Type{T}) where {T <: Period}
+    nf = time_per_time(Nanosecond, T)
+    return CompoundPeriod(Nanosecond(ceil(x * nf)))
```

Comment:
```suggestion
function CompoundPeriod(x::Real, ::Type{T}) where {T <: Period}
    return CompoundPeriod(Nanoseconds(ceil(value(Nanosecond(T(1)) * nf)))
```

I don't really know if `ceil` is expected there or not in counting the fractional number of Nanoseconds. Since if not it seems to get much simpler (but fails in converti f the multiplication with `nf` doesn't result in an exact number of Nanoseconds)

```
CompoundPeriod(x::Real, ::Type{T}) where {T <: Period} = CompoundPeriod(Nanosecond(T(1)) * nf)
```

Reply:
This suggestion (with `Nanoseconds` -> `Nanosecond`) fails the tests, so something is wrong with the suggested logic.

Reply:
Ah, I see what you mean. In other words, first compute `n::Int` nanoseconds from the input, then apply an exact convert. I don't think non-exact conversions are supported, which is basically what this PR is doing.

Alternatively, we could define something like `nearest_nanosecond(::Real, ::Period)`, and then use that to define this functionality.

Reply:
That seemed to be what `ceil(x * time_per_time())` was computing, though I didn't check tests to see what it would do on those

Reply:
`time_per_time` basically duplicates part of the logic in `define_conversions` below, so I assume that it's somewhat irreducible. I'm not sure if we can somehow leverage that code, we only need `N` from the conversion.

Reply:
What makes `time_per_time` different from `value(convert(Nanosecond, T(1)))` then, since the logic seems to be duplicated?

Reply:
`time_per_time` cannot error, since it returns the number of time units when going from one conversion to another (and that's possible because it's only defined when going from larger periods to smaller periods, and there's no possibility for overflow since it's independent of the value we're working with). As for `value(convert(Nanosecond, T(1)))`, there's no guarantee for this to be exact, and therefore errors in the converts.

We could change the converts below to use `time_per_time`, and define a single convert method instead of eval-ing a bunch:

```
function Base.convert(::Type{T}, x::Tc) where {T <: Period, Tc <: Period}
	N = T === Period ? 1 : time_per_time_new(T, Tc)
	vmax = typemax(Int64) ÷ N
	vmin = typemin(Int64) ÷ N
	vmin ≤ value(x) ≤ vmax || throw(InexactError(:convert, T, x))
	return T(value(x) * N)
end
```
Would this be preferable?

We could apply a similar pattern to extract `N` for the "`# more-precise periods`" conversions / promotions, but I don't know what benefit this might have.

Reply:
It seemed like `time_per_time` was only being called as `nanoseconds_per_time`, and the rest of it was unused and so untested for this PR?

Reply:
Ah, yes, that's correct, should we rename it and only define those methods that provide the number of nanoseconds?

Reply:
Sure, but then how does that differ from asking for the count of nanoseconds in 1 unit of `T` as: `value(convert(Nanosecond, T(1)))`

Reply:
Ahhhhh, I see. I was just confused by your suggestion, which had a trivial typo. `CompoundPeriod(Nanoseconds(ceil(value(Nanosecond(T(1)) * nf)))` should be `CompoundPeriod(Nanoseconds(ceil(x * value(Nanosecond(T(1)) * nf)))`. That's much simpler. Updating...

Reply:
Oh, yeah, sorry. I had originally had `nf` separated like that when testing, and then tried to simplify it more to post